### ISSUE TITLE
[graphql-ruby-client] Critical dependency: the request of a dependency is an expression

### ISSUE THREAD

**Issue by in15** [rmosolgo/graphql-ruby#3607]
**Describe the bug**

We are utilizing graphql-ruby-client in our frontend React SPA. Packages are installed with npm and we would like to configure this with Apollo & Pusher.

When generating a build or attempting to run the local dev server, we encounter the following warning:
```
project: Failed to compile.
project: ./node_modules/graphql-ruby-client/sync/prepareRelay.js
project: Critical dependency: the request of a dependency is an expression
project:  @ ./node_modules/graphql-ruby-client/sync/generateClient.js
project:  @ ./node_modules/graphql-ruby-client/index.js

```
**Versions**

`graphql` version: 15.5.2
`graphql-ruby-client`: 1.10.0

**GraphQL schema**

N/A

**GraphQL query**
`N/A`

**Steps to reproduce**
Installed the package with `npm install` to our frontend project
Started our local dev server

**Expected behavior**
No issues or warnings

**Actual behavior**
copied/pasted from output snippet above
```
project: Failed to compile.
project: ./node_modules/graphql-ruby-client/sync/prepareRelay.js
project: Critical dependency: the request of a dependency is an expression
project:  @ ./node_modules/graphql-ruby-client/sync/generateClient.js
project:  @ ./node_modules/graphql-ruby-client/index.js

```

**Additional context**

We use webpack 4.x

We are utilizing Apolllo-Client, and we are on the latest version of that. Our thinking though, is that this warning we are seeing is not a result of integration with other libraries. Maybe (hopefully?) we're wrong.

**Comment by rmosolgo** [comment]
How are you requiring `graphql-ruby-client` in your project? Could you please share any lines of code that load files from this project? (I expect the fix to be modifing those `import`/`require` statements -- I see `index.js` is in the error message, but you shouldn't use `index.js` for web front-end projects because it leads to this problem!)

**Comment by in15** [comment]
Hi @rmosolgo good question. 

We have the latest `graphql-ruby-client` in our package.json
```
"graphql-ruby-client": "^1.10.0"
```

And we are attempting to utilize PusherLink by importing in like so
```
import PusherLink from 'graphql-ruby-client/subscriptions/PusherLink'
...
const pusherLink = new PusherLink({ pusher: pusherClient })

...
const link = ApolloLink.from([pusherLink, httpLink])
  
return new ApolloClient({
    cache,
    link,
  });
```

**Comment by xxx** [comment]
We also use Apollo and Pusher, and also get this warning, but it doesn't cause compilation failures in our case.

example trace:

```
13:56:37 webpack-dev-server.1 | WARNING in ./node_modules/graphql-ruby-client/sync/prepareRelay.js 41:22-47
13:56:37 webpack-dev-server.1 | Critical dependency: the request of a dependency is an expression
13:56:37 webpack-dev-server.1 |     at CommonJsRequireContextDependency.getWarnings (/home/mpd/git-sources/catchandrelease/node_modules/webpack/lib/dependencies/ContextDependency.js:40:18)
13:56:37 webpack-dev-server.1 |     at Compilation.reportDependencyErrorsAndWarnings (/home/mpd/git-sources/catchandrelease/node_modules/webpack/lib/Compilation.js:1454:24)
13:56:37 webpack-dev-server.1 |     at /home/mpd/git-sources/catchandrelease/node_modules/webpack/lib/Compilation.js:1258:10
13:56:37 webpack-dev-server.1 |     at AsyncSeriesHook.eval [as callAsync] (eval at create (/home/mpd/git-sources/catchandrelease/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:13:1)
13:56:37 webpack-dev-server.1 |     at AsyncSeriesHook.lazyCompileHook (/home/mpd/git-sources/catchandrelease/node_modules/tapable/lib/Hook.js:154:20)
13:56:37 webpack-dev-server.1 |     at Compilation.finish (/home/mpd/git-sources/catchandrelease/node_modules/webpack/lib/Compilation.js:1253:28)
13:56:37 webpack-dev-server.1 |     at /home/mpd/git-sources/catchandrelease/node_modules/webpack/lib/Compiler.js:672:17
13:56:37 webpack-dev-server.1 |     at _done (eval at create (/home/mpd/git-sources/catchandrelease/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:7:1)
13:56:37 webpack-dev-server.1 |     at eval (eval at create (/home/mpd/git-sources/catchandrelease/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:126:22)
13:56:37 webpack-dev-server.1 |     at /home/mpd/git-sources/catchandrelease/node_modules/webpack/lib/Compilation.js:1185:12
13:56:37 webpack-dev-server.1 | ℹ ｢wdm｣: Compiled with warnings
```

**Comment by rmosolgo** [comment]
Thanks for sharing those details ... I just don't understand, why does webpack care about this file that isn't used by the project 😖 

@in15, is there anything else in the backtrace besides the 5 lines posted above? I'm curious if there are any clues hidden in it...

**Comment by chriswright47** [comment]
Thanks so much for helping look at this, @rmosolgo! I'm on @in15's team and I've been trying to figure out why webpack cares about that `prepareRelay` file since we don't reference it or import it anywhere. Here is a slightly more complete output from our Travis build. In local development, we get the same warning but the app still works as expected (and the `PusherLink` also works). We have our production build set to fail on any warnings though, hence the "Failed to compile" message from Travis. I'm investigating our build script now to see if I can get some more insight.
```
scheduler: > node scripts/build.js
scheduler: Creating an optimized production build...
scheduler: Uploaded static/js/runtime.09baa495.js.map to Rollbar
scheduler: Uploaded static/js/vendor.583518f2.chunk.js.map to Rollbar
scheduler: Uploaded static/js/vendors~app.e7e82bbf.chunk.js.map to Rollbar
scheduler: Uploaded static/js/vendors~vendor.d4d9617f.chunk.js.map to Rollbar
scheduler: Uploaded static/js/app.d429f482.chunk.js.map to Rollbar
scheduler: Uploaded static/js/vendors~app~vendor.5d844b81.chunk.js.map to Rollbar
scheduler: Treating warnings as errors because process.env.CI = true.
scheduler: Most CI servers set it automatically.
scheduler: Failed to compile.
scheduler: ./node_modules/graphql-ruby-client/sync/prepareRelay.js
scheduler: Critical dependency: the request of a dependency is an expression
scheduler:  @ ./node_modules/graphql-ruby-client/sync/generateClient.js
scheduler:  @ ./node_modules/graphql-ruby-client/index.js
scheduler:  @ multi @apollo/client abl-reporting apollo-boost apollo-link-context auth0-lock axios bluebird graphql graphql-ruby-client lodash pluralize prop-types pusher-js query-string raf react react-autocomplete react-bootstrap react-datepicker react-dom react-redux react-relay react-router-dom react-transition-group react-virtualized recharts redux redux-thunk request request-promise reselect rollbar uuid
```

**Comment by chriswright47** [comment]
Update: I got our build to work by explicitly ignoring the directory that this `prepareRelay` file is in in our webpack.config file useing the [noParse](https://webpack.js.org/configuration/module/#modulenoparse) option. With this addition, our travis build passes and the app still works as expected
```
module.exports = {
  ...
  module: {
    noParse: /graphql-ruby-client\/sync/,
    ...
  }
}
```

**Comment by xxx** [comment]
@rmosolgo Serendipitously, my teammate @pvenable recently tweaked our configuration with the following:

![Screenshot from 2021-09-07 16-37-56](https://user-images.githubusercontent.com/41678/132423528-e88e9ae3-3eed-4abe-adfa-fa983850b4c1.png)

and we now compile clean, without any warnings at all.

**Comment by hueter** [comment]
This thread was helpful; I also had compilation issues and realized it was from using the old import style. I must have missed it in the changelog. For me, to fix things I changed: 

```tsx
import { ActionCableLink } from "graphql-ruby-client";
```

to 

```tsx
import ActionCableLink from "graphql-ruby-client/subscriptions/ActionCableLink";
```

For anyone reading who is confused, the explanation here is that the `index.ts` is an [index / "barrel" file](https://basarat.gitbook.io/typescript/main-1/barrel) that exports everything. So Webpack will actually read through all of the files that are imported from `index.ts` when you do `import { something } from "graphql-ruby-client";` (unless you have some fancy tree-shaking set up). This import pattern will break things in this library because `pusher-js` and `relay-runtime`, etc. are all "optional dependencies", so webpack will see `import` statements for all of those dependencies and not find them, since you probably only have 1 of them installed. Hence you have to point directly to the `/subscriptions/ActionCableLink` file so that you only import what you need. Notably, the other way you can fix this "module not found" issue is by installing every optional dependency -- not recommended since it will increase your bundle size.

Not sure if there is a use-case I'm missing, but it might be worth deleting [these lines in `index.ts`](https://github.com/rmosolgo/graphql-ruby/blob/master/javascript_client/src/index.ts#L12-L14)  if it causes build failures due to missing optional dependencies. It would be a "breaking change" but it would prevent build compilation errors. I'm happy to help with a PR and update the docs if that's something that would help.

I also mentioned a [pipe dream fix a while ago](https://github.com/rmosolgo/graphql-ruby/issues/2802) to do scoped packages for this repo (similar to `@apollo-client/`) but that is a lot of refactoring / restructuring 😅

**Comment by rmosolgo** [comment]
Thanks for sharing that clarification, @hueter. 

> (unless you have some fancy tree-shaking set up)

I made the mistake of assuming that tree-shaking was the default, so I updated the code and docs accordingly. When people started reporting errors like this one, I reverted the changes to the docs... but I didn't change the old code, since it _would_ work for anyone with tree-shaking. But it meant that these errors continued on 😖
